### DISCUSSION TITLE
Does wsl-vpnkit is mandatory for Podman desktop on windows to connect outside network ?

### DISCUSSION THREAD

**Discussion by Bharath-vi** [podman-desktop/podman-desktop#3151] [General]
### Is your enhancement related to a problem? Please describe

Do we need to use wsl-vpnkit for Podman in corporate network ?

Without wsl-vpnkit podman is not connecting with internet/quay/dockerhub.

### Describe the solution you'd like

podman should work without wsl-vpnkit

**Comment by benoitf** [comment]
hello @Bharath-vi in the upcoming Podman 4.6 there will be a new user network option
https://github.com/containers/podman/pull/18303

I think it should solve your problem

**Reply by Bharath-vi** [reply]
Thank you for the response @benoitf

**Comment by Bharath-vi** [comment]
Hi @benoitf 
I have enabled user mode networking.
I getting this error after restarting my sysytem, do we need to add these file everytime ?

cp: cannot stat '/mnt/wsl/resolv.conf': No such file or directory
Error: error setting up user-mode networking: exit status 1



PS C:\Users> podman machine start
Starting machine "podman-machine-default"
Starting user-mode networking...
cp: cannot stat '/mnt/wsl/resolv.conf': No such file or directory
Error: error setting up user-mode networking: exit status 1

then, i have added resolv.conf and wsl.conf
resolv.conf
nameserver 1.1.1.1
nameserver 8.8.8.8
wsl.conf
[network]
generateResolvConf = false

**Reply by benoitf** [reply]
cc @n1hility

**Comment by n1hility** [comment]
@Bharath-vi you are likely running an older generation of WSL, `wsl --update` should pull a later version. When this is updated `wsl --version` should return version info. If its just help text then you need to apply all the latest win updates, then rerun `wsl --update` After that is working you should be able to recreate the machine with user-mode-networking enabled

**Reply by Bharath-vi** [reply]
Hi @n1hility ,
User mode networking is working as expected,
PS C:\Users\AL36413> wsl -l -v
NAME                      STATE           VERSION 
* wsl-vpnkit                Stopped         2
  podman-net-usermode       Running         2 
  podman-machine-default    Running         2  

Only isssue is resolv.conf file we have to create after restarting system, Is there any option we can add the resolv.conf file permanently.

**Reply by n1hility** [reply]
Hey @Bharath-vi , glad that is working for you! With user-mode networking, you should not change/edit resolv.conf; it's generated as part of the user-mode networking setup, and is essential to DNS traffic flowing the path of the Windows host. Changing could mean DNS queries would not resolve with a VPN.

**Comment by slemeur** [comment]
cc @themr0c ! This is a nice one for the troubleshooting guide

**Comment by themr0c** [comment]
As I undestand it now, it looks more like a procedure "Using Podman in a network behind a user-controlled VPN":

Ideal case: In Podman Desktop, when initializing the Podman machine, select the **User mode networking** option.

But if I am not mistaken, we do not have this option. Do we plan to add it?

Current situation:

(Can someone help me test it, I don't have a Windows host at hands at the moment):

# Accessing with Podman a network behind a user-controlled VPN on Windows

When you require Podman to access resources that are behind a user-controlled VPN, enable the user mode networking in your Podman machine.

#### Prerequisites

- Windows host with updated WSL2.
- On the host, the user controls the VPN activation.
- An running Podman machine *<podman_machine_name>*.

#### Procedure

1. Enable user mode networking:

   ```shell-session
   $  podman machine set --user-mode-networking *<podman_machine_name>*
   ```

2. SSH to the Podman machine

   ```
   $ podman machine ssh *<podman_machine_name>*
   ```

3. Create and edit the `/etc/resolv.conf/` file:

   ```
   nameserver 1.1.1.1
   nameserver 8.8.8.8
   ```

4. Create and edit the `/mnt/wsl/wsl.conf` file:
   
   ```
   [network]
   generateResolvConf = false
   ```

5. Exit the Podman machine SSH session.

   ```shell-session
   $ exit
   ```

6. Restart the Podman machine.


#### Verification

- You can access resources that are behind the VPN.

**Reply by n1hility** [reply]
@themr0c Thanks for proposing this. Although steps 2-5 **should not** be done, and will introduce problems. I would add one step at the top, a prerequisite, which is to be sure you are running the current generation of WSL (wsl --version should print a version number, if it prints help text an update process is required).  I will paste some instructions on that in one sec.

**Reply by n1hility** [reply]
@themr0c here you go, its a bit rough, but hopefully useful! Also BTW I am planning to do a fix to detect and print better guidance when the WSL version is too old.

### Windows / WSL Prerequisite

Run `wsl --verrsion` and confirm you are running WSL2 of a version at least 1.2.5.0 or greater. If you get a bunch of help text with this command you are running the previous generation.

If you are not running the current version, follow the following steps:

1. First confirm you are running the appropriate version of Windows with all current updates applied. Run the `winver` command and check the release number. For Windows 10, you need to be on 21H1, 21H2, or 22H2 on Windows 10, and for Windows 11 you need to be on at least 21H2. If you are not running an older release of windows, follow the Windows update process and will eventually suggest  an update to one of these versions. If you are running an enterprise/corporate build of Windows, Windows Update may be disabled by your IT team. In which case, contact them for steps to update. Note that multiple reboots may be required after each stage of updates before you are suggested to install the newer update pack.
2. Once you are running a recent enough version of Windows, run windows update one more time to be sure no further updates are suggested. Even if you are running the appropriate version, the WSL update mechanism was introduced after their initial release, so this step is important.
3. Run the `wsl --update`, which should then convert your WSL installation to the current store based version. If you receive an error, and your system is a corporate laptop, your IT team may have disabled usage of the Windows store (provides this stream). If that's the case you can run `wsl --update --web-download` to fetch the latest version from Github directly instead. If you still receive an error, your IT administrator may have disabled "Windows Store Applications", which is necessary to run WSL. Contact them to get this capability enabled on your system.
4. After `wsl --update` completes, reboot your system, and you should see the expected version output from `wsl --version` on your next login.

## Additional info relating to podman machine set --user-mode-networking

If you have an earlier podman machine instance on your system (before 4.6.0), you will receive a notification from the set command that you need to recreate your machine off of the latest image stream, which contains the necessary support binaries for user-mode networking. To do so following the following steps:

1. Backup any container data you have on your existing podman machine as this will be lost. One way to do this is with the `podman export command` which you can run on either the WSL prompt (wsl -d podman-machine-default), or via the ssh prompt `podman machine ssh`
2. Stop the podman machine instance `podman machine stop`
3. Destroy the podman machine instance `podman machine rm`. Answer yes to prompts.
4. Create a new instance with the user-mode networking option `podman machine init --user-mode-networking`, and any other parameters you need
5. Start your podman machine instance `podman machine start`

**Reply by lucasjl** [reply]
Hello @n1hility. I follow all steps recommended but then I get to this error message after 'podman machine start':

Error: error setting up user-mode networking: exit status 42

While there are issues about this error (https://github.com/containers/podman/issues/20982) I don't see any responses yet as it's very recent.

Thanks!

**Reply by benoitf** [reply]
@lucasjl there is a temporary failure in the user mode networking now.

you could use workaround that I paste on https://github.com/containers/podman/issues/20982#issuecomment-1852838422 

or use podman-desktop airgap installers (they embed the file so it's not downloading the latest one)
